modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java

```diff
@@ -69,205 +53,69 @@ protected GridCacheDistributedQueryFuture(GridCacheContext<K, V> ctx, long reqId
-
-        return futFinish;
+    @Override protected CacheQueryReducer<R> reducer() {
+        return reducer;
     }
 
-    /** {@inheritDoc} */
-    @Override protected void loadPage() {
-        assert !Thread.holdsLock(this);
-
-        Collection<ClusterNode> nodes = null;
-
-        synchronized (this) {
-            if (!isDone() && rcvd.containsAll(subgrid)) {
-                rcvd.clear();
-
-                nodes = nodes();
-            }
-        }
-
-        if (nodes != null)
-            cctx.queries().loadPage(reqId, qry.query(), nodes, false);
+    /** Set reducer. */
+    void reducer(DistributedCacheQueryReducer<R> reducer) {
+        this.reducer = reducer;
     }
 
     /** {@inheritDoc} */
-    @Override protected void loadAllPages() throws IgniteInterruptedCheckedException {
-        assert !Thread.holdsLock(this);
-
-        U.await(firstPageLatch);
-
-        Collection<ClusterNode> nodes = null;
-
-        synchronized (this) {
-            if (!isDone() && !subgrid.isEmpty())
-                nodes = nodes();
-        }
-
-        if (nodes != null)
-            cctx.queries().loadPage(reqId, qry.query(), nodes, true);
-    }
-
-    /**
-     * @return Nodes to send requests to.
-     */
-    private Collection<ClusterNode> nodes() {
-        assert Thread.holdsLock(this);
-
-        Collection<ClusterNode> nodes = new ArrayList<>(subgrid.size());
-
-        for (UUID nodeId : subgrid) {
-            ClusterNode node = cctx.discovery().node(nodeId);
-
-            if (node != null)
-                nodes.add(node);
-        }
-
-        return nodes;
+    @Override public Collection<R> get() throws IgniteCheckedException {
+        return get0();
     }
 
     /** {@inheritDoc} */
-    @Override public boolean onDone(Collection<R> res, Throwable err) {
-        boolean done = super.onDone(res, err);
-
-        // Must release the lath after onDone() in order for a waiting thread to see an exception, if any.
-        firstPageLatch.countDown();
-
-        return done;
+    @Override public Collection<R> get(long timeout, TimeUnit unit) throws IgniteCheckedException {
+        return get0();
     }
 
     /** {@inheritDoc} */
-    @Override public boolean onCancelled() {
-        firstPageLatch.countDown();
-
-        return super.onCancelled();
+    @Override public Collection<R> getUninterruptibly() throws IgniteCheckedException {
+        return get0();
     }
 
-    /** {@inheritDoc} */
-    @Override public void onTimeout() {
-        firstPageLatch.countDown();
+    /**
+     * Completion of distributed query future depends on user that iterates over query result with lazy page loading.
+     * Then {@link #get()} can lock on unpredictably long period of time. So we should avoid call it.
+     */
+    private Collection<R> get0() throws IgniteCheckedException {
+        if (!isDone())
+            throw new IgniteIllegalStateException("Unexpected lock on iterator over distributed cache query result.");
```

Comment:
Why do you allow to get result via get() method, but forbid to wait?
If noone must call get() then it should throws exception in anyway.
If the method is never called and you don't care then overriding is useless.

Reply:
The only place we use get() it's in this class, some lines above.

https://github.com/apache/ignite/blob/0add93b488ebe7265b2a0f951fcdfd55adbcc01f/modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheDistributedQueryFuture.java#L82

But I agree, that it a little bit misleading. So I will replace this line of code with `super.get()` and then we can forbid `get()` at all.